libs/experimental/langchain_experimental/data_anonymizer/deanonymizer_matching_strategies.py

```diff
@@ -15,3 +18,73 @@ def default_matching_strategy(text: str, deanonymizer_mapping: MappingDataType)
         for anonymized, original in deanonymizer_mapping[entity_type].items():
             text = text.replace(anonymized, original)
     return text
+
+
+def case_insensitive_matching_strategy(
+    text: str, deanonymizer_mapping: MappingDataType
+) -> str:
+    """
+    Case insensitive matching strategy for deanonymization.
+    It replaces all the anonymized entities with the original ones
+        irrespective of their letter case.
+
+    Args:
+        text: text to deanonymize
+        deanonymizer_mapping: mapping between anonymized entities and original ones
+    """
+
+    # Iterate over all the entities (PERSON, EMAIL_ADDRESS, etc.)
+    for entity_type in deanonymizer_mapping:
+        for anonymized, original in deanonymizer_mapping[entity_type].items():
+            # Use regular expressions for case-insensitive matching and replacing
+            text = re.sub(anonymized, original, text, flags=re.IGNORECASE)
+    return text
+
+
+def ngram_fuzzy_matching_strategy(
+    text: str, deanonymizer_mapping: MappingDataType, fuzzy_threshold: int = 80
+) -> str:
+    """
+    N-gram fuzzy matching strategy for deanonymization.
+    It replaces all the anonymized entities with the original ones.
+    It uses fuzzy matching to find the position of the anonymized entity in the text.
+    It generates n-grams of the same length as the anonymized entity from the text and
+    uses fuzzy matching to find the position of the anonymized entity in the text.
```

Comment:
please consider if n-grams of the same length is enough? I can imagine situations when llm adds or removes a part of anonymized entity. Maybe we should compare (_n-1_, _n_, _n+1_)-grams ?

Reply:
I added option to check varying length of ngrams :smile: